### ISSUE TITLE
LoRA on FlashAttention-based model fails due to float16/bfloat16 assertion error?

### ISSUE THREAD

**Issue by patricks-lab** [Dao-AILab/flash-attention#676]
Hi guys - I'm using an older version of FlashAttention (~0.0.1) since a custom model I'm using requires it. In this case, I'm trying to train using LoRA on the Wqkv attention linear layers, but it seems to be complaining about some type of error that complains that the "qkv" matrix is not of type float16/bfloat16. 

Here's my LoRA setup, in which case you can see that I've added a LoRA adapter on all the 'Wqkv' linear layers, and froze all other layers (Note the custom model I'm using works perfectly fine without the LoRA freezing below):
```
target_mods = []
for n, m in model.named_modules():
    if ('Wqkv' in n):
        target_mods += [n]
        print(n, type(m)) 

from peft import LoraConfig
from peft import get_peft_model
config = LoraConfig(
    target_modules=target_mods,
)
model = get_peft_model(model, config)
model.print_trainable_parameters()
model.to(device)
```

Here's what a Wqkv layer looks like in my setup:
```
(self_attn): FlashMHA(
          (Wqkv): Linear(in_features=512, out_features=1536, bias=True)
          (inner_attn): FlashAttention()
          (out_proj): Linear(in_features=512, out_features=512, bias=True)
        )
```

I've tried replacing all aspects of the Wqkv matrix to torch.float16 as mentioned in the error, but this doesn't fix the error at all:
```
for i in range(12):
    model.transformer_encoder.layers[i].self_attn.Wqkv.lora_A.default = model.transformer_encoder.layers[i].self_attn.Wqkv.lora_A.default.to(torch.float16)
    model.transformer_encoder.layers[i].self_attn.Wqkv.lora_B.default = model.transformer_encoder.layers[i].self_attn.Wqkv.lora_B.default.to(torch.float16)
    model.transformer_encoder.layers[i].self_attn.Wqkv.lora_dropout.default = model.transformer_encoder.layers[i].self_attn.Wqkv.lora_dropout.default.to(torch.float16)
    model.transformer_encoder.layers[i].self_attn.Wqkv = model.transformer_encoder.layers[i].self_attn.Wqkv.to(torch.float16)
    model.transformer_encoder.layers[i].self_attn.out_proj = model.transformer_encoder.layers[i].self_attn.out_proj.to(torch.float16)
```

Here's a part of the error I'm persistently getting - which happens during training:
```
File model.py:710, in FlashTransformerEncoderLayer.forward(self, src, src_mask, src_key_padding_mask, **kwargs)
    708     src = src + self.dropout2(src2)
    709 else:
--> 710     src2 = self.self_attn(src, key_padding_mask=src_key_padding_mask_)[0]
    711     src = src + self.dropout1(src2)
    712     src = self.norm1(src)

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/torch/nn/modules/module.py:1190, in Module._call_impl(self, *input, **kwargs)
   1186 # If we don't have any hooks, we want to skip the rest of the logic in
   1187 # this function, and just call forward.
   1188 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1189         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1190     return forward_call(*input, **kwargs)
   1191 # Do not call functions when jit is used
   1192 full_backward_hooks, non_full_backward_hooks = [], []

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/flash_attn/flash_attention.py:99, in FlashMHA.forward(self, x, key_padding_mask, need_weights)
     97 qkv = self.Wqkv(x)
     98 qkv = rearrange(qkv, 'b s (three h d) -> b s three h d', three=3, h=self.num_heads)
---> 99 context, attn_weights = self.inner_attn(qkv, key_padding_mask=key_padding_mask,
    100                                         need_weights=need_weights, causal=self.causal)
    101 return self.out_proj(rearrange(context, 'b s h d -> b s (h d)')), attn_weights

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/torch/nn/modules/module.py:1190, in Module._call_impl(self, *input, **kwargs)
   1186 # If we don't have any hooks, we want to skip the rest of the logic in
   1187 # this function, and just call forward.
   1188 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1189         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1190     return forward_call(*input, **kwargs)
   1191 # Do not call functions when jit is used
   1192 full_backward_hooks, non_full_backward_hooks = [], []

File ~/miniconda3/envs/myenv/lib/python3.9/site-packages/flash_attn/flash_attention.py:36, in FlashAttention.forward(self, qkv, key_padding_mask, causal, cu_seqlens, max_s, need_weights)
     28 """Implements the multihead softmax attention.
     29 Arguments
     30 ---------
   (...)
     33     key_padding_mask: a bool tensor of shape (B, S)
     34 """
     35 assert not need_weights
---> 36 assert qkv.dtype in [torch.float16, torch.bfloat16]
     37 assert qkv.is_cuda
     39 if cu_seqlens is None:

AssertionError: 
```

TLDR - Would appreciate any tips on how to get LoRA to work with custom FlashAttention-based model, which doesn't work since it's complaining the qkv matrix has to be of type float16 or bfloat16. 

Thanks in advance!
-Patrick
